processor/k8sprocessor/processor.go

```diff
@@ -122,119 +70,83 @@ func (kp *kubernetesprocessor) Shutdown(context.Context) error {
-			podIP = kp.k8sIPFromAttributes(resource.Attributes())
-		}
-
-		// Check if the receiver detected client IP.
-		if podIP == "" {
-			if c, ok := client.FromContext(ctx); ok {
-				podIP = c.IP
-			}
-		}
-
-		if podIP != "" {
-			if resource.IsNil() {
-				resource.InitEmpty()
-			}
-			resource.Attributes().InsertString(k8sIPLabelName, podIP)
-		}
+		_ = kp.processResource(ctx, rs.Resource(), k8sIPFromAttributes())
+	}
 
-		// Don't invoke any k8s client functionality in passthrough mode.
-		// Just tag the IP and forward the batch.
-		if kp.passthroughMode {
-			continue
-		}
+	return td, nil
+}
 
-		// add k8s tags to resource
-		attrsToAdd := kp.getAttributesForPodIP(podIP)
-		if len(attrsToAdd) == 0 {
+// ProcessMetrics process metrics and add k8s metadata using resource IP, hostname or incoming IP as pod origin.
+func (kp *kubernetesprocessor) ProcessMetrics(ctx context.Context, md pdata.Metrics) (pdata.Metrics, error) {
+	rm := md.ResourceMetrics()
+	for i := 0; i < rm.Len(); i++ {
+		ms := rm.At(i)
+		if ms.IsNil() {
 			continue
 		}
 
-		if resource.IsNil() {
-			resource.InitEmpty()
-		}
-
-		attrs := resource.Attributes()
-		for k, v := range attrsToAdd {
-			attrs.InsertString(k, v)
-		}
+		_ = kp.processResource(ctx, ms.Resource(), k8sIPFromAttributes(), k8sIPFromHostnameAttributes())
 	}
 
-	return kp.nextTraceConsumer.ConsumeTraces(ctx, td)
+	return md, nil
 }
 
-// ConsumeMetrics process metrics and add k8s metadata using resource hostname as pod origin.
-// TODO: Move to internal data model once it's available in contrib.
-func (kp *kubernetesprocessor) ConsumeMetrics(ctx context.Context, metrics pdata.Metrics) error {
-	mds := internaldata.MetricsToOC(metrics)
-
-	for i := range mds {
-		md := &mds[i]
-		var podIP string
-
-		// Check if a collector/agent or a prior processor has already annotated the metrics with IP.
-		if md.Resource.GetLabels() != nil {
-			podIP = md.Resource.Labels[k8sIPLabelName]
-		}
-
-		// Most of the metric receivers uses "host.hostname" resource label (which is represented as
-		// Node.Identifier.HostName in OpenCensus format) to identify metrics origin.
-		// In k8s environment, it's set to a pod IP address. If the value doesn't represent
-		// an IP address, we skip it.
-		if podIP == "" && md.Node.GetIdentifier().GetHostName() != "" {
-			hostname := md.Node.Identifier.HostName
-			if net.ParseIP(hostname) != nil {
-				podIP = hostname
+func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pdata.Resource, attributeExtractors ...ipExtractor) error {
+	var podIP string
 
-				// Set pod IP as resource label.
-				if md.Resource == nil {
-					md.Resource = &resourcepb.Resource{}
-				}
-				if md.Resource.Labels == nil {
-					md.Resource.Labels = map[string]string{}
-				}
-				md.Resource.Labels[k8sIPLabelName] = podIP
+	if !resource.IsNil() {
+		for _, extractor := range attributeExtractors {
+			if podIP == "" {
+				podIP = extractor(resource.Attributes())
```

Comment:
Will the last extractor always win or do we stop if one of the extractors find an IP?

Reply:
This should stop at the first extractor that retrieves the IP

Reply:
Perhaps an explicit `break` statement would make the intent more clear?